src/rdmnet/disc/lightweight/lwmdns_common.c

```diff
@@ -150,48 +180,478 @@ const uint8_t* lwmdns_parse_resource_record(const uint8_t*     buf_begin,
+{
+  if (!buf_begin || !label || !str_buf)
+    return false;
+
+  DomainNameLabel label_data = DOMAIN_NAME_LABEL_INIT;
+  if (get_domain_name_label(buf_begin, label, &label_data))
+  {
+    memcpy(str_buf, label_data.label, label_data.length);
+    str_buf[label_data.length] = '\0';
+    return true;
+  }
+  return false;
+}
+
+txt_record_parse_result_t lwmdns_txt_record_to_broker_info(const uint8_t*    txt_data,
+                                                           uint16_t          txt_data_len,
+                                                           DiscoveredBroker* db)
 {
-  /*
-  bool data_changed = false;
+  txt_keys_found_mask_t keys_found = 0;
+  bool                  data_changed = false;
 
   const uint8_t* cur_ptr = txt_data;
+  bool           parsed_txt_vers = false;
   while (cur_ptr - txt_data < txt_data_len)
   {
     uint8_t txt_len = *cur_ptr++;
     if ((cur_ptr + txt_len - txt_data) > txt_data_len)
-      return data_changed;
+      break;
+
+    TxtRecordItemRef item;
+    item.key = cur_ptr;
+    bool found_equals = false;
 
     for (const uint8_t* data = cur_ptr; data < cur_ptr + txt_len; ++data)
     {
       if ((char)*data == '=')
       {
+        found_equals = true;
+        item.key_len = (uint8_t)(data - cur_ptr);
+        item.value = data + 1;
+        item.value_len = txt_len - (item.key_len + 1);
+        if (item.key_len == 0 || item.key_len + 1 > txt_len)
+          break;
+
+        if (parsed_txt_vers)
+        {
+          if (parse_txt_item(&item, db, &keys_found))
+            data_changed = true;
+          break;
+        }
+        else
+        {
+          // The TxtVers key must be the first key in the TXT record.
+          if (parse_txt_vers(&item))
+          {
+            parsed_txt_vers = true;
+            break;
+          }
+          else
+          {
+            return kTxtRecordParseError;
+          }
+        }
       }
     }
+    // An additional TXT item with no '=' (key only)
+    if (!found_equals && txt_len > 0)
+    {
+      item.key_len = txt_len;
+      item.value = NULL;
+      item.value_len = 0;
+      if (parse_txt_item(&item, db, &keys_found))
+        data_changed = true;
+    }
+    cur_ptr += txt_len;
+  }
+
+  if (ALL_TXT_KEYS_FOUND(keys_found))
+  {
+    return (data_changed ? kTxtRecordParseOkDataChanged : kTxtRecordParseOkNoDataChanged);
+  }
+  else
+  {
+    return kTxtRecordParseError;
+  }
+}
+
+bool parse_txt_vers(const TxtRecordItemRef* item)
+{
+  if (item->key_len != sizeof(E133_TXT_VERS_KEY) - 1)
+    return false;
+
+  if (memcmp(item->key, E133_TXT_VERS_KEY, sizeof(E133_TXT_VERS_KEY) - 1) != 0)
+    return false;
+
+  if (binary_atoi(item->value, item->value_len) != E133_DNSSD_TXTVERS)
+    return false;
+
+  return true;
+}
+
+bool parse_txt_item(const TxtRecordItemRef* item, DiscoveredBroker* db, txt_keys_found_mask_t* found_mask)
+{
+  // E133Scope
+  if (item->key_len == sizeof(E133_TXT_SCOPE_KEY) - 1 &&
```

Comment:
I'm seeing a lot of this pattern here, and in the previous function.  Perhaps a utility function that takes two keys & and a size?